why-cant-we.md

```diff
@@ -0,0 +1,286 @@
+## What we've already achieved
+
+Before going over ideas that probably won't fly, it's worth reminding ourselves
+of the improvements we've already made over the last decade of typed Python
+development:
+
+- [PEP 526](https://peps.python.org/pep-0526/) (Python 3.6): Replaced type
+  comments for variable annotations with native syntax.
+- [PEP 585](https://peps.python.org/pep-0585/) (Python 3.9): Added subscripting
+  support to built-in and library types, removing the need for `typing.List` and
+  similar objects.
+- [PEP 604](https://peps.python.org/pep-0604/) (Python 3.10): Support for using
+  `|` to create union types.
+- [PEP 673](https://peps.python.org/pep-0673/) (Python 3.11): Added `typing.Self`,
+  replacing a common complex pattern.
+- [PEP 695](https://peps.python.org/pep-0695/) (Python 3.12): Added concise, native 
+  syntax for generics, removing the need for manual variance declarations.
+- [PEP 563](https://peps.python.org/pep-0563/), [PEP 649](https://peps.python.org/pep-0649/),
+  and [PEP 749](https://peps.python.org/pep-0749/) went through a complex history,
+  but the upshot for static typing users is that they can avoid worrying about
+  runtime errors in type annotations.
+
+## Dictionary literals as inline TypedDicts
+
+### Idea
+
+Allow writing a `TypedDict` type inline as a dictionary literal. Instead of:
+
+```python
+from typing import TypedDict
+
+class Movie(TypedDict):
+    name: str
+    year: int
+
+def print_movie(movie: Movie) -> None:
+    print(f"{movie['name']} came out in {movie['year']}")
+```
+
+Write:
+
+```python
+def print_movie(movie: {"name": str, "year": int}) -> None:
+    print(f"{movie['name']} came out in {movie['year']}")
+```
+
+Or in a type alias:
+
+```python
+type Movie = {"name": str, "year": int}
+```
+
+As a bonus, this also allows using keys that are not valid identifiers,
+like the existing call-based syntax for creating a TypedDict.
+
+The `NotRequired` and `ReadOnly` type qualifiers could also be used,
+e.g. `{"name": str, "year": NotRequired[int]}`.
+
+Note that some of the below problems can be mitigated by wrapping
+the dictionary, e.g. `TypedDict[{"name": str}]`. The in-preparation
+PEP 764 will propose this syntax.
+
+### Problems
+
+_`|` operator_: The `|` operator is already defined for dictionaries,
+and its behavior doesn't match the expected behavior for unions in
+the type system:
+
+```pycon
+>>> {"a": int} | {"b": str}
+{'a': <class 'int'>, 'b': <class 'str'>}
+>>> {"a": int} | {"a": str}
+{'a': <class 'str'>}
+```
+
+This would make it difficult to introspect annotations using unions
+of inline TypedDicts at runtime.
+
+Creating a union of an inline TypedDict and a type object currently
+throws an error:
+
+```pycon
+>>> {"a": int} | int
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: unsupported operand type(s) for |: 'dict' and 'type'
+```
+
+However, this is fixable by changing the implementation of `__or__`
+on type objects.
```

Comment:
a bit nitpicky, but the specific example you give above this paragraph would actually have to be solved with a change to `type.__ror__` rather than `type.__or__`:

```suggestion
However, this is fixable by changing the implementation of `__or__`
and `__ror__` on type objects.
```

Reply:
To be even more nitpicky, those are actually the same :) They both map to the `nb_or` slot.